xds/internal/resolver/xds_resolver.go

```diff
@@ -236,94 +284,248 @@ func (r *xdsResolver) sendNewServiceConfig(cs *configSelector) bool {
-	// error cases, and the fields might not be set when the error happens.
-	if r.cancelWatch != nil {
-		r.cancelWatch()
+	r.onResolutionComplete()
+}
+
+func (r *xdsResolver) onListenerResourceUpdate(update xdsresource.ListenerUpdate) {
+	r.currentListener = update
+	r.listenerUpdateRecvd = true
+
+	if update.InlineRouteConfig != nil {
+		// If there was a previous route config watcher because of a non-inline
+		// route configuration, cancel it.
+		r.rdsResourceName = ""
+		if r.routeConfigWatcher != nil {
+			r.routeConfigWatcher.stop()
+			r.routeConfigWatcher = nil
+		}
+
+		r.applyRouteConfigUpdate(*update.InlineRouteConfig)
+		return
 	}
-	if r.xdsClientClose != nil {
-		r.xdsClientClose()
+
+	// We get here only if there was no inline route configuration.
+
+	// If the route config name has not changed, send an update with existing
+	// route configuration and the newly received listener configuration.
+	if r.rdsResourceName == update.RouteConfigName {
+		r.onResolutionComplete()
+		return
 	}
-	r.closed.Fire()
-	r.logger.Infof("Shutdown")
+
+	// If the route config name has changed, cancel the old watcher and start a
+	// new one. At this point, since we have not yet resolved the new route
+	// config name, we don't have to send an update to the channel.
+	r.rdsResourceName = update.RouteConfigName
+	if r.routeConfigWatcher != nil {
+		r.routeConfigWatcher.stop()
+		r.currentVirtualHost = nil
+		r.routeConfigUpdateRecvd = false
+	}
+	r.routeConfigWatcher = newRouteConfigWatcher(r.rdsResourceName, r)
+}
+
+// onError propagates the error up to the channel. And since this is invoked
+// only for non-resource-not-found error, we don't have to update resolver state
+// and we can keep using the old config.
+func (r *xdsResolver) onError(err error) {
+	r.cc.ReportError(err)
+}
+
+// Contains common functionality to be executed when resources of either type
+// are removed.
+func (r *xdsResolver) onResourceNotFoundCommon() {
+	// We cannot remove clusters from the service config that have ongoing RPCs.
+	// Instead, what we can do is to send an erroring (nil) config selector
+	// along with normal service config. This will ensure that new RPCs will
+	// fail, and once the active RPCs complete, the reference counts on the
+	// clusters will come down to zero. At that point, we will send an empty
+	// service config config with no addresses. This results in the pick-first
+	// LB policy being configured on the channel, and since there are no
+	// address, pick-first will put the channel in TRANSIENT_FAILURE.
+	r.sendNewServiceConfig(nil)
+
+	// Stop and dereference the active config selector, if one exists.
+	r.curConfigSelector.stop()
+	r.curConfigSelector = nil
+}
+
+func (r *xdsResolver) onListenerResourceNotFound() {
+	if r.routeConfigWatcher != nil {
+		r.routeConfigWatcher.stop()
+		r.routeConfigWatcher = nil
+		r.currentVirtualHost = nil
+		r.routeConfigUpdateRecvd = false
+	}
```

Comment:
Move this block after 508, and only stop route config watcher and stop() to avoid duplicating code.